### ISSUE TITLE
interrupted repair by reboot and segfault occurred on another db node

### ISSUE THREAD

**Issue by amoskong** [scylladb/scylla#6543]
*Installation details*
Scylla version (or git commit hash): 4.1.rc1-0.20200528.bb8fcbff68f
Cluster size: 6
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-00907d577946117dd (eu-west-1)

### Description

In repair_streaming_err, repair was running on target node db-node-3, hard reboot db-node-3 to interrupt the repair task. but segfault occurred on another node db-nod-7.

#### Cluster info:
- Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-1 [54.154.213.73 | 10.0.181.251]
- Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-2 [34.254.229.44 | 10.0.223.182]
- Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-3 [54.217.54.156 | 10.0.145.190]
- Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-4 [34.246.186.37 | 10.0.156.205]
- Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-5 [34.246.177.143 | 10.0.188.223]
- Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-6 [34.243.48.156 | 10.0.5.206]
- Node longevity-cdc-100gb-4h-4-1-loader-node-89aa4120-1 [54.75.54.41 | 10.0.201.135]
- Node longevity-cdc-100gb-4h-4-1-loader-node-89aa4120-2 [3.250.107.79 | 10.0.122.19]
- Node longevity-cdc-100gb-4h-4-1-monitor-node-89aa4120-1 [63.33.210.5 | 10.0.147.191]
- Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 [52.215.236.147 | 10.0.53.9]
- Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-8 [54.195.153.147 | 10.0.30.83]

#### Test Info
- test id: 89aa4120-9057-4d7f-921d-3a9f330b99da
- db log: https://cloudius-jenkins-test.s3.amazonaws.com/89aa4120-9057-4d7f-921d-3a9f330b99da/20200530_052410/db-cluster-89aa4120.zip
- sct log: https://cloudius-jenkins-test.s3.amazonaws.com/89aa4120-9057-4d7f-921d-3a9f330b99da/20200530_052410/sct-runner-89aa4120.zip
- job link: https://jenkins.scylladb.com/job/scylla-4.1/job/longevity/job/longevity-cdc-100gb-4h/2/

#### db-node-7 was added in last second nemesis: grow_shrink_cluster
- added db-node-89aa4120-7, and it's up (ready)
- decommission db-node-89aa4120-6

```
2020-05-30 02:11:40,187 >>>>>>>>>>>>>Started random_disrupt_method grow_shrink_cluster
2020-05-30 02:31:38,130 Cluster longevity-cdc-100gb-4h-4-1-db-cluster-89aa4120 (AMI: ['ami-00907d577946117dd'] Type: i3.4xlarge): (1/1) nodes ready, node Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 [52.215.236.147 | 10.0.53.9] (seed: False). Time elapsed: 1090 s
2020-05-30 02:56:54,940 Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-6 [34.243.48.156 | 10.0.5.206] (seed:False): Command 'nodetool  decommission ' duration -> 1192.7787644863129 s
2020-05-30 02:32:01,993 sdcm.nemesis.ChaosMonkey: New node added longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7
2020-05-30 02:32:01,994 sdcm.nemesis.ChaosMonkey: Finish cluster grow
2020-05-30 02:57:06,195  <<<<<<<<<<<<<Finished random_disrupt_method grow_shrink_cluster

```

##### Segfault in last Nemesis: repair_streaming_err
- repair on on db-node-3
- hard reboot on db-node-3 to interrupt the repair
- then `Segfault` occurred on db-node-7

```
2020-05-30 03:05:23,794  >>>>>>>>>>>>>Started random_disrupt_method repair_streaming_err
Current Target: Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-3 [54.217.54.156 | 10.0.145.190]

2020-05-30 03:08:44,786 > Wait for: Wait for streaming starts: timeout - 200 seconds - expired
2020-05-30 03:08:44,787 > sdcm.nemesis.ChaosMonkey: wait for random between 10s to 10m

2020-05-30 03:09:01,877 RemoteCmdRunner [centos@10.0.145.190]: Running command "nodetool  repair "...

2020-05-30 03:17:14,883 > sdcm.nemesis.ChaosMonkey: Interrupt the task (repair) to trigger streaming error
2020-05-30 03:17:14,883 > sdcm.nemesis.ChaosMonkey: Interrupt the task by hard reboot

2020-05-30 03:17:14,974 > Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-3 [54.217.54.156 | 10.0.145.190] (seed: False): Hardly rebooting node
2020-05-30 03:17:15,557 > sdcm.nemesis.ChaosMonkey: repair is stopped

2020-05-30T03:17:20+00:00  longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7
| scylla: Segmentation fault on shard 0.
| scylla: Backtrace:
| scylla: 0x0000000002e0d002
| scylla: 0x0000000002db6644
| scylla: 0x0000000002db6945
| scylla: 0x0000000002db6990
| scylla: 0x00007feb3d61ab1f
| scylla: 0x0000000000c64a2f
| scylla: 0x0000000003250ee9
| scylla: 0x0000000002442137
| scylla: 0x0000000002db2b70
| scylla: 0x0000000002db2d7f
| scylla: 0x0000000002de7b95
| scylla: 0x0000000002d52419
| scylla: 0x0000000002d534ce
| scylla: 0x0000000000d24ae6
| scylla: /opt/scylladb/libreloc/libc.so.6+0x00000000000271a2
| scylla: 0x0000000000c4a02d
```

### Executed Nemesis:

```
2020-05-30 00:38:01,509  >>>>>>>>>>>>>Started random_disrupt_method no_corrupt_repair
2020-05-30 00:39:07,068  <<<<<<<<<<<<<Finished random_disrupt_method no_corrupt_repair
2020-05-30 00:47:37,554  >>>>>>>>>>>>>Started random_disrupt_method restart_then_repair_node
2020-05-30 00:55:19,885  <<<<<<<<<<<<<Finished random_disrupt_method restart_then_repair_node
2020-05-30 01:03:27,065  >>>>>>>>>>>>>Started random_disrupt_method network_reject_node_exporter
2020-05-30 01:08:27,486  <<<<<<<<<<<<<Finished random_disrupt_method network_reject_node_exporter
2020-05-30 01:17:03,194  >>>>>>>>>>>>>Started random_disrupt_method network_reject_thrift
2020-05-30 01:17:13,565  <<<<<<<<<<<<<Finished random_disrupt_method network_reject_thrift
2020-05-30 01:25:43,836  >>>>>>>>>>>>>Started random_disrupt_method mgmt_backup
2020-05-30 01:29:04,349  >>>>>>>>>>>>>Started random_disrupt_method no_corrupt_repair
2020-05-30 01:51:12,897  <<<<<<<<<<<<<Finished random_disrupt_method no_corrupt_repair
2020-05-30 01:58:31,441  >>>>>>>>>>>>>Started random_disrupt_method hard_reboot_node
2020-05-30 02:03:13,373  <<<<<<<<<<<<<Finished random_disrupt_method hard_reboot_node
2020-05-30 02:11:40,187  >>>>>>>>>>>>>Started random_disrupt_method grow_shrink_cluster
< t:2020-05-30 02:56:54,940 f:cluster.py      l:2430 c:sdcm.cluster_aws     p:DEBUG > Node longevity-cdc-100gb-4h-4-1-db-node-89aa4120-6 [34.243.48.156 | 10.0.5.206] (seed:False): Command 'nodetool  decommission ' duration -> 1192.7787644863129 s
2020-05-30 02:57:06,195  <<<<<<<<<<<<<Finished random_disrupt_method grow_shrink_cluster
2020-05-30 03:05:23,794  >>>>>>>>>>>>>Started random_disrupt_method repair_streaming_err
```

**Comment by amoskong** [comment]
```
$ addr2line -Cfpi -e /usr/lib/debug/opt/scylladb/libexec/scylla-4.1.rc1-0.20200528.bb8fcbff68f.x86_64.debug 
0x0000000002e0d002
0x0000000002db6644
0x0000000002db6945
0x0000000002db6990
0x00007feb3d61ab1f
0x0000000000c64a2f
0x0000000003250ee9
0x0000000002442137
0x0000000002db2b70
0x0000000002db2d7f
0x0000000002de7b95
0x0000000002d52419
0x0000000002d534ce
0x0000000000d24ae6
/opt/scylladb/libreloc/libc.so.6+0x00000000000271a2
0x0000000000c4a02d
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /usr/include/fmt/format.h:1158
seastar::backtrace_buffer::append_backtrace() at /usr/include/fmt/format.h:1158
 (inlined by) print_with_backtrace at ./build/release/seastar/./seastar/src/core/reactor.cc:762
seastar::print_with_backtrace(char const*) at /usr/include/fmt/format.h:1158
sigsegv_action at /usr/include/fmt/format.h:1158
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3455
 (inlined by) _FUN at ./build/release/seastar/./seastar/src/core/reactor.cc:3451
?? ??:0
std::_Function_handler<void (utils::updateable_value_base*), utils::updateable_value_source<std::unordered_map<seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::basic_sstring<char, unsigned int, 15u, true>, std::hash<seastar::basic_sstring<char, unsigned int, 15u, true> >, std::equal_to<seastar::basic_sstring<char, unsigned int, 15u, true> >, std::allocator<std::pair<seastar::basic_sstring<char, unsigned int, 15u, true> const, seastar::basic_sstring<char, unsigned int, 15u, true> > > > >::for_each_ref(std::function<void (utils::updateable_value<std::unordered_map<seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::basic_sstring<char, unsigned int, 15u, true>, std::hash<seastar::basic_sstring<char, unsigned int, 15u, true> >, std::equal_to<seastar::basic_sstring<char, unsigned int, 15u, true> >, std::allocator<std::pair<seastar::basic_sstring<char, unsigned int, 15u, true> const, seastar::basic_sstring<char, unsigned int, 15u, true> > > > >*)>)::{lambda(utils::updateable_value_base*)#1}>::_M_invoke(std::_Any_data const&, utils::updateable_value_base*&&) at main.cc:?
operator() at /usr/include/fmt/format.h:1158
 (inlined by) seastar::logger::really_do_log(seastar::log_level, char const*, seastar::logger::stringer const*, unsigned long) at ./build/release/seastar/./seastar/src/util/log.cc:246
_ZZN7seastar6logger6do_logIJRKNS_13basic_sstringIcjLj15ELb1EEES5_RKNSt15__exception_ptr13exception_ptrEEEEvNS_9log_levelEPKcDpOT_ENKUlDpSD_E_clIJNS0_8stringerESJ_SJ_EEEDaSG_ at /usr/include/fmt/format.h:1158
 (inlined by) void seastar::logger::do_log<seastar::basic_sstring<char, unsigned int, 15u, true> const&, seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::__exception_ptr::exception_ptr const&>(seastar::log_level, char const*, seastar::basic_sstring<char, unsigned int, 15u, true> const&, seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::__exception_ptr::exception_ptr const&) at ././seastar/include/seastar/util/log.hh:329
 (inlined by) void seastar::logger::log<seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::basic_sstring<char, unsigned int, 15u, true>, std::__exception_ptr::exception_ptr>(seastar::log_level, char const*, seastar::basic_sstring<char, unsigned int, 15u, true> const&, seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::__exception_ptr::exception_ptr const&) at ././seastar/include/seastar/util/log.hh:125
 (inlined by) void seastar::logger::warn<seastar::basic_sstring<char, unsigned int, 15u, true> const&, seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::__exception_ptr::exception_ptr&>(char const*, seastar::basic_sstring<char, unsigned int, 15u, true> const&, seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::__exception_ptr::exception_ptr&) at ././seastar/include/seastar/util/log.hh:150
 (inlined by) repair_writer::create_writer(seastar::sharded<database>&, unsigned int)::{lambda(std::__exception_ptr::exception_ptr)#4}::operator()(std::__exception_ptr::exception_ptr) const at ./repair/row_level.cc:532
 (inlined by) seastar::future<> seastar::futurize<seastar::future<> >::invoke<repair_writer::create_writer(seastar::sharded<database>&, unsigned int)::{lambda(std::__exception_ptr::exception_ptr)#4}&, std::__exception_ptr::exception_ptr>(repair_writer::create_writer(seastar::sharded<database>&, unsigned int)::{lambda(std::__exception_ptr::exception_ptr)#4}&, std::__exception_ptr::exception_ptr&&) at ././seastar/include/seastar/core/future.hh:1748
 (inlined by) auto seastar::futurize_invoke<repair_writer::create_writer(seastar::sharded<database>&, unsigned int)::{lambda(std::__exception_ptr::exception_ptr)#4}&, std::__exception_ptr::exception_ptr>(repair_writer::create_writer(seastar::sharded<database>&, unsigned int)::{lambda(std::__exception_ptr::exception_ptr)#4}&, std::__exception_ptr::exception_ptr&&) at ././seastar/include/seastar/core/future.hh:1832
 (inlined by) _ZZN7seastar6futureIJEE16handle_exceptionIZN13repair_writer13create_writerERNS_7shardedI8databaseEEjEUlNSt15__exception_ptr13exception_ptrEE2_EES1_OT_ENUlSC_E_clIS1_EES1_SC_ at ././seastar/include/seastar/core/future.hh:1518
 (inlined by) _ZZZZN7seastar6futureIJEE17then_wrapped_nrvoIS1_ZNS1_16handle_exceptionIZN13repair_writer13create_writerERNS_7shardedI8databaseEEjEUlNSt15__exception_ptr13exception_ptrEE2_EES1_OT_EUlSD_E_EENS_8futurizeISC_E4typeEOT0_ENKUlvE_clEvENUlRNS_8internal22promise_base_with_typeIJEEEONS_12future_stateIJEEEE_clESO_SR_ENKUlvE_clEv at ././seastar/include/seastar/core/future.hh:1344
 (inlined by) _ZN7seastar8futurizeINS_6futureIJEEEE22satisfy_with_result_ofIZZZNS2_17then_wrapped_nrvoIS2_ZNS2_16handle_exceptionIZN13repair_writer13create_writerERNS_7shardedI8databaseEEjEUlNSt15__exception_ptr13exception_ptrEE2_EES2_OT_EUlSG_E_EENS0_ISF_E4typeEOT0_ENKUlvE_clEvENUlRNS_8internal22promise_base_with_typeIJEEEONS_12future_stateIJEEEE_clESQ_ST_EUlvE_EEvOSP_SG_ at ././seastar/include/seastar/core/future.hh:1741
 (inlined by) _ZZZN7seastar6futureIJEE17then_wrapped_nrvoIS1_ZNS1_16handle_exceptionIZN13repair_writer13create_writerERNS_7shardedI8databaseEEjEUlNSt15__exception_ptr13exception_ptrEE2_EES1_OT_EUlSD_E_EENS_8futurizeISC_E4typeEOT0_ENKUlvE_clEvENUlRNS_8internal22promise_base_with_typeIJEEEONS_12future_stateIJEEEE_clESO_SR_ at ././seastar/include/seastar/core/future.hh:1343
 (inlined by) _ZN7seastar12continuationINS_8internal22promise_base_with_typeIJEEEZZNS_6futureIJEE17then_wrapped_nrvoIS5_ZNS5_16handle_exceptionIZN13repair_writer13create_writerERNS_7shardedI8databaseEEjEUlNSt15__exception_ptr13exception_ptrEE2_EES5_OT_EUlSH_E_EENS_8futurizeISG_E4typeEOT0_ENKUlvE_clEvEUlRS3_ONS_12future_stateIJEEEE_JEE15run_and_disposeEv at ././seastar/include/seastar/core/future.hh:504
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/include/fmt/format.h:1158
seastar::reactor::run_some_tasks() at /usr/include/fmt/format.h:1158
seastar::reactor::run_some_tasks() at /usr/include/fmt/format.h:1158
 (inlined by) seastar::reactor::run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2721
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at /usr/include/boost/program_options/variables_map.hpp:146
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at /usr/include/boost/program_options/variables_map.hpp:146
main at ./main.cc:483
?? ??:0
_start at ??:?
```

**Comment by amoskong** [comment]
Hi @slivne @bhalevy , can you help to check if this issue is related with or same as https://github.com/scylladb/scylla/issues/6370 ?

**Comment by bhalevy** [comment]
@amoskong this looks completely different.
The segfault happens in a call to `rlogger.warn`
while in #6370 the segfault happens in `seastar::expiring_fifo`

Is there a coredump?

Cc @asias

**Comment by slivne** [comment]
@amoskong can you please reference the uploaded coredump

**Comment by slivne** [comment]
2 nodes down causes availability issues

**Comment by amoskong** [comment]
> @amoskong this looks completely different.
> The segfault happens in a call to `rlogger.warn`
> while in #6370 the segfault happens in `seastar::expiring_fifo`
> 
> Is there a coredump?
> 
> Cc @asias

No coredump was collected.  /CC @slivne

**Comment by roydahan** [comment]
This happened on a longevity with CDC.
@amoskong let's run reproducer of this nemesis only with a simple longevity without CDC.
(It may reduce the severity of the issue because CDC isn't GA for 4.1)

**Comment by slivne** [comment]
@asias can you check the params passed to the log call -

**Comment by amoskong** [comment]
> This happened on a longevity with CDC.
> @amoskong let's run reproducer of this nemesis only with a simple longevity without CDC.
> (It may reduce the severity of the issue because CDC isn't GA for 4.1)

The reproduce job is running : 
- https://jenkins.scylladb.com/job/scylla-4.1/job/longevity/job/Reproducer/job/longevity-cdc-4h-test-only-repair/2/console
- only with Corrupt_data_then_repair nemesis
- CDC disabled.

**Comment by asias** [comment]
> @asias can you check the params passed to the log call -

rlogger.warn("repair_writer: keyspace={}, table={}, write_end_of_stream failed: {}", _schema->ks_name(), _schema->cf_name(), ep);

Probably the repair_writer object was removed when the above runs. 

My theory is:

node3 runs repair as  master, node4 is follower
node4 starts the repair writer
node3 runs hard reboot
node4 notices node3 is down by gossip
node4 runs the on_dead gossip handler which removes the repair_meta object which in turn removes the repiair_writer object
node4 runs the exception handler which calls the logger
node4 has after free and triggers the segfault


```
17192 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]:  [shard 0] gossip - InetAddress 10.0.145.190 is now DOWN, status = shutdown
17193 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: Segmentation fault on shard 0.
17194 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: Backtrace:
17195 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002e0d002
17196 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002db6644
17197 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002db6945
17198 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002db6990
17199 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x00007feb3d61ab1f
17200 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000000c64a2f
17201 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000003250ee9 
17202 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002442137
17203 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002db2b70
17204 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002db2d7f
17205 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002de7b95
17206 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002d52419
17207 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000002d534ce
17208 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000000d24ae6
17209 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: /opt/scylladb/libreloc/libc.so.6+0x00000000000271a2
17210 May 30 03:17:20 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 scylla[2377]: 0x0000000000c4a02d
17211 May 30 03:18:38 longevity-cdc-100gb-4h-4-1-db-node-89aa4120-7 systemd-coredump[37612]: Process 2377 (scylla) of user 996 dumped core.
```

**Comment by slivne** [comment]
@asias any progress on this one ?

**Comment by asias** [comment]
My use after free theory above should not happen. We always wait for the repair_writer to finish before we free the repair_meta object in the gossip  on_dead handler. I added debug code and ran repair locally to verify.

**Comment by amoskong** [comment]
I didn't reproduced this segfault after disabled CDC.

- job: https://jenkins.scylladb.com/job/scylla-4.1/job/longevity/job/Reproducer/job/longevity-cdc-4h-test-only-repair/3/
- CDC disabled.

**Comment by slivne** [comment]
Since CDC is experimental in 4.1 - this issue will not be a blocker for 4.1 (and will be a blocker in the release CDC is GAd).

**Comment by asias** [comment]
We have bunch of repair related fix merged in master and 4.2 branch. Dow we hit the issue again with 4.2 + CDC enabled?

**Comment by asias** [comment]
I believe this is fixed by: repair: Fix race between create_writer and wait_for_writer_done (e6f640441a726c9b6f9ccfb2a482d809ad962041). I am closing  it.
